zebra-consensus/src/script.rs

```diff
@@ -12,11 +12,181 @@
+
+impl<ZS> tower::Service<Request> for Verifier<ZS>
+where
+    ZS: tower::Service<zebra_state::Request, Response = zebra_state::Response, Error = BoxError>,
+    ZS::Future: Send + 'static,
+{
+    type Response = ();
+    type Error = BoxError;
+    type Future =
+        Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;
+
+    fn poll_ready(
+        &mut self,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), Self::Error>> {
+        self.state.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request) -> Self::Future {
+        use futures_util::FutureExt;
+
+        let input = &req.transaction.inputs()[req.input_index];
+
+        match input {
+            transparent::Input::PrevOut { outpoint, .. } => {
+                let output = self.state.call(zebra_state::Request::AwaitUtxo(*outpoint));
+                let transaction = req.transaction;
+                let branch_id = self.branch;
+                let input_index = req.input_index;
+
+                async move {
+                    let previous_output = match output.await? {
+                        zebra_state::Response::Utxo(output) => output,
+                        _ => unreachable!("AwaitUtxo always responds with Utxo"),
+                    };
+
+                    zebra_script::is_valid(
+                        transaction,
+                        branch_id,
+                        (input_index as u32, previous_output),
+                    )?;
+
+                    Ok(())
+                }
+                .boxed()
+            }
+            transparent::Input::Coinbase { .. } => unimplemented!(
+                "how should we handle verifying coinbase transactions in the script::Verifier?"
+            ),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::sync::Arc;
+
+    use color_eyre::Report;
+
+    use zebra_chain::{
+        block::Block, parameters::Network, parameters::NetworkUpgrade,
+        serialization::ZcashDeserializeInto, transparent,
+    };
+    use zebra_test::transcript::{TransError, Transcript};
+
+    #[tokio::test]
+    async fn happy_path() -> Result<(), Report> {
+        happy_path_test().await
+    }
+
+    #[spandoc::spandoc]
+    async fn happy_path_test() -> Result<(), Report> {
+        zebra_test::init();
+        let mut inputs = zebra_test::vectors::MAINNET_BLOCKS
+            .range(1..10)
+            .flat_map(|(_, block_bytes)| {
+                let block: Block = block_bytes.zcash_deserialize_into().unwrap();
+                block.transactions.into_iter()
+            })
+            .map(|transaction| {
+                transaction
+                    .inputs()
+                    .iter()
+                    .cloned()
+                    .collect::<Vec<transparent::Input>>()
+            })
+            .flat_map(|inputs| inputs.into_iter());
+
+        assert!(inputs.any(|input| matches!(input, transparent::Input::PrevOut { .. })));
```

Comment:
I tried writing a test but it ended up being perfectly, beautifully useless. This assert fails. Not really sure how to test this one...

Reply:
The first block with a non-coinbase transaction is at least block 101, because of the coinbase maturity rule.

Should we focus on getting the sync acceptance tests working? (#1141)

If we only want to test finalized state, then we could add a sync test for the second checkpoint, which has 2000 blocks.

Reply:
I think that might be necessary, but even then I'm not sure we can test this code meaningfully via the acceptance test, given the restriction only being able to chain verify post sapling blocks.